returnn/datasets/distrib_files.py

```diff
@@ -283,11 +287,14 @@ def init_seq_order(self, epoch: Optional[int] = None, seq_list=None, seq_order=N
             else:
                 raise ValueError(f"{self}: seq_ordering {self.seq_ordering!r} not supported")
             file_bins = self._distribute_evenly_by_size(
-                num_bins=self._num_shards * self.partition_epoch,
+                num_bins=self._num_data_shards * self._num_file_shards * self.partition_epoch,
```

Comment:
I wonder a bit about this. Does it make sense actually to have these two types of sharding? Isn't that the same? I think we should just reuse the existing sharding. I.e. undo all the changes in `DistribFilesDataset`, just rename `_num_shards` to `_num_data_shards`. Or why do you actually call it `_num_data_shards` at all? I initially suggested to just call it `_num_shards`. I think that's a better name. Then `DistribFilesDataset` does not need any changes.

Reply:
It is a better name, I made the distinction to have this conflict-free wrt. DistribFilesDataset's own implementation of sharding. But yeah maybe this should just be unified, in the same way MultiProc sets the shards for the wrapped ones DistribFilesDataset can fetch them from the distributed context.

Reply:
I don't see the point of this distinction. I think specifically we don't want that these are separate types of shardings. These are the same type of sharding. Maybe we can actually generalize it a bit, and let this be an option in RETURNN, similar to `horovod_dataset_distribution`, e.g. just `distrib_dataset_sharding = True` or so, and if that is enabled, it would pass rank/size as the sharding opts for the top level train dataset.

Reply:
I have an implementation for this generalistic approach now, let's discuss it in a follow-up PR. Can you review this PR as-is again to see if we can merge it now? :)